In [12]:
import pandas as pd 
import numpy as np 
import os 

In [13]:
# creating a function
def add_num(x,y):
 sum = x + y
 return(sum)

In [14]:
add_num(8, 9)


17

In [15]:
# importing the order_prod_merge

df_ord_prod = pd.read_pickle(r"C:\Users\okumb\Downloads\instaCart_Bucket_04_2024\Data\Prepared_Data\orders_products_merged.pkl")

In [16]:
df_ord_prod.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [17]:
df = df_ord_prod[: 1000000]

In [18]:
df.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [19]:
# creating a function to categorise the product prices

def price_label(row):
    
    if row['prices'] <= 5:
        return "Low range price"
    elif (row['prices'] >5) and (row['prices']<= 15):
        return "Medium range price"
    elif row['prices'] >15:
        return "High range price"
    else:
        return "Not enough data"

In [20]:
df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\okumb\AppData\Local\Temp/ipykernel_2292/935325922.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [21]:
df['price_range'].value_counts()

Medium range price    756450
Low range price       243550
Name: price_range, dtype: int64

In [22]:
df['prices'].max()

14.8

In [23]:
df_ord_prod.loc[df_ord_prod['prices'] > 16, 'price_range_loc'] = 'high_price_range' 

In [24]:
df_ord_prod.loc[(df_ord_prod['prices'] > 5) & (df_ord_prod['prices'] <= 15), 'price_range_loc'] ='medium_price_range' 

In [25]:
df_ord_prod.loc[df_ord_prod['prices'] <= 5, 'price_range_loc'] = 'low_price_range' 

In [26]:
df_ord_prod['price_range_loc'].value_counts()

medium_price_range    21860860
low_price_range       10126321
high_price_range        391761
Name: price_range_loc, dtype: int64

In [27]:
# counting the days of the week and sort it in ascending order

df_ord_prod['order_days_of_week'].sort_values(ascending = True).value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_days_of_week, dtype: int64

In [28]:
# allocating the busiest day to 0 and the least busy to day 4

result = []

for value in df_ord_prod['order_days_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy day')
    else: 
        result.append('Regular busy day')

In [29]:
result

['Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Least busy day',
 'Least busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Least busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Least busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Least busy day',
 'Least busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Regular busy day',
 'Re

In [30]:
df_ord_prod['busiest_day'] = result

In [31]:
df_ord_prod['busiest_day'].value_counts(dropna = False)

Regular busy day    22416875
Busiest day          6204182
Least busy day       3783802
Name: busiest_day, dtype: int64

In [32]:
# renaming the busiest day rows from days order days of the week column as well the slowest day

df_ord_prod['new_busiest_day'] = result

df_ord_prod['new_busiest_day'].head()



0    Regular busy day
1    Regular busy day
2    Regular busy day
3      Least busy day
4      Least busy day
Name: new_busiest_day, dtype: object

In [33]:
df_ord_prod['new_busiest_day'].head()

0    Regular busy day
1    Regular busy day
2    Regular busy day
3      Least busy day
4      Least busy day
Name: new_busiest_day, dtype: object

In [34]:
df_ord_prod.loc[(df_ord_prod['order_days_of_week'] == 0) | (df_ord_prod['order_days_of_week'] == 1), 'new_busiest_day'] = 'Busiest days'



In [35]:
df_ord_prod.loc[(df_ord_prod['order_days_of_week'] == 3) | (df_ord_prod['order_days_of_week'] == 4), 'new_busiest_day'] = 'Slowest days'

In [36]:
df_ord_prod.loc[df_ord_prod['order_days_of_week'] == 6, 'new_busiest_day'] = 'Regular busy days'
df_ord_prod.loc[df_ord_prod['order_days_of_week'] == 2, 'new_busiest_day'] = 'Regular busy days'
df_ord_prod.loc[df_ord_prod['order_days_of_week'] == 5, 'new_busiest_day'] = 'Regular busy days'

In [37]:

df_ord_prod['new_busiest_day'].replace('Regular busy day', 'Regular busy days', inplace=True)


In [38]:
df_ord_prod['new_busiest_day']

0           Regular busy days
1                Slowest days
2                Slowest days
3                Slowest days
4                Slowest days
                  ...        
32404854    Regular busy days
32404855    Regular busy days
32404856    Regular busy days
32404857         Slowest days
32404858         Slowest days
Name: new_busiest_day, Length: 32404859, dtype: object

In [39]:
df_ord_prod['new_busiest_day'].value_counts(dropna = False)

Regular busy days    12916111
Busiest days         11864412
Slowest days          7624336
Name: new_busiest_day, dtype: int64

### Check the values of this new column for accuracy. Note any observations in markdown format.

##### Regular busy day    22416875
##### Busiest day          6204182
##### Least busy day       3783802
##### Name: busiest_day, dtype: int64


##### Regular busy days    12916111
##### Busiest days         11864412
##### Slowest days          7624336
##### Name: new_busiest_day, dtype: int64



## you could see from the new information, busiest days and slowest days have more numbers to them than the pervious one.

In [40]:
df_ord_prod.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge     price_range_loc  \
0         Soda        77              7     9.0   both  medium_price_range   
1         Soda        77              7     9.0   both  medium_price_range   
2         Soda        77              7     9.0   both  medium_price_range   
3         Soda        77              7     9.0   both  medium_price_range   
4         Soda        77              7     9.0   both  medium_price_range   

        busiest_day    new_busiest_day  
0  Regular busy day  Regular busy days  
1  Regular busy day       Slowest days  
2  Regular busy day       Slowest days  
3    Least busy day       Slowest days  
4    Least busy day       Slowest days

#### When too many users make Instacart orders at the same time, the app freezes. 
#### The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, 
#### they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” 
#### Create a new column containing these labels called “busiest_period_of_day.”

In [41]:
df_ord_prod['order_hour_of_day'].sort_values(ascending = True).value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [42]:
df_ord_prod.loc[df_ord_prod['order_hour_of_day'].isin([10,11,14,15,13,12,16,9,17]), 'busiest_period_of_day'] = 'Most orders' 


In [43]:
df_ord_prod.loc[df_ord_prod['order_hour_of_day'].isin([8,18,19,20,7,21,22,23,6,0,1]), 'busiest_period_of_day'] = 'Average orders'

In [44]:
df_ord_prod.loc[df_ord_prod['order_hour_of_day'].isin([5,2,4,3]), 'busiest_period_of_day'] = 'Fewest orders'

#### Print the frequency for this new column.

In [45]:
df_ord_prod['busiest_period_of_day'].value_counts(dropna = False).sort_values(ascending = False)

Most orders       23205725
Average orders     8937275
Fewest orders       261859
Name: busiest_period_of_day, dtype: int64

In [46]:
path = r"C:\Users\okumb\Downloads\instaCart_Bucket_04_2024\Data\Prepared_Data"

In [47]:
# export the df_ord_prod as a pickle file

df_ord_prod.to_pickle(os.path.join(path, 'orders_products_hours.pkl'))





In [48]:
df_ord_prod.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge     price_range_loc  \
0         Soda        77              7     9.0   both  medium_price_range   
1         Soda        77              7     9.0   both  medium_price_range   
2         Soda        77              7     9.0   both  medium_price_range   
3         Soda        77              7     9.0   both  medium_price_range   
4         Soda        77              7     9.0   both  medium_price_range   

        busiest_day    new_busiest_day busiest_period_of_day  
0  Regular busy day  Regular busy days        Average orders  
1  Regular busy day       Slowest days        Average orders  
2  Regular busy day       Slowest days           Most orders  
3    Least busy day       Slowest days        Average orders  
4    Least busy day       Slowest days           Most orders